### ISSUE TITLE
Compatible with browserify / commonjs modules?

### ISSUE THREAD

**Issue by alasdaircr** [almasaeed2010/AdminLTE#567]
Hi -

I've spent a few hours attempting to get AdminLTE loading alongside my other JS modules which are bundled using browserify.

I'm a bit lost now, I don't know what modifications are needed to AdminLTE in order to support this, what configuration to browserify is needed, and whether I need to require('adminlte') somewhere in my app.js file.

**Comment by almasaeed2010** [comment]
I don't use Browserify, so I am not sure if I can help.

**Comment by Maximization** [comment]
@alasdaircr Admin LTE won't work with browserify right out of the box. You need to shim it first. Assuming you have already downloaded from npm:

* Install browserify-shim as a dev dependency
```
npm install browserify-shim --save-dev
```
* Add aliases for the required dependencies in your `package.json` (optional, but recommended)
```
"browser": {
    "jquery": "./node_modules/admin-lte/plugins/jQuery/jQuery-2.1.4.min.js",
    "bootstrap": "./node_modules/admin-lte/bootstrap/js/bootstrap.min.js",
    "admin-lte": "./node_modules/admin-lte/dist/js/app.min.js"
  }
```
* Let browserify know we're shimming it
```
"browserify": {
    "transform": ["browserify-shim"]
  }
```
* Attach jquery as $ global var and explicitly state Admin LTE's dependencies
```
"browserify-shim": {
    "jquery": "$",
    "admin-lte": {
      "depends": ["jquery:$","bootstrap"]
    }
  }
```
* Finally you can `require('admin-lte')` in your `main.js` file and using the command bellow in the terminal or as an npm script browserify will do the res
```
"scripts": {
    "build": "browserify -d main.js -o bundle.js"
  }
```

This will give you the basic functionality enough to get you up and running. For any additional plugins that Admin LTE depends on you will have to add them to your `package.json` file. Let me know how this works out for you and shoot if you have any questions.

**Comment by drbarto** [comment]
Is this solution still supposed to work? I gave up my [attempts](https://github.com/almasaeed2010/AdminLTE/issues/1055) to make AdminLTE work with webpack and moved on to browserify; unfortunately, the same issue occurs -- when doing `require('admin-lte)` an Exception (`Error: AdminLTE requires jQuery`) is raised.

I followed your guide step by step, so please can you tell me this appraoch is still working for you?

(P.S.: using AdminLTE 2.3.3 and browserify 13.0.1)

**Comment by Maximization** [comment]
I just checked with updated packages (AdminLTE 2.3.3, browserify 13.0.1 and browserify-shim 3.8.12) and surely it still works. Can you paste the contents of your package.json?

**Comment by drbarto** [comment]
Hey, thanks for checking back! Here's my package.json:

    "devDependencies": {
       "browserify-shim": "^3.5.0"
    },
    "dependencies": {
      "admin-lte": "^2.3.3",
      "backbone": "^1.3.3",
      "bootstrap": "^3.3.6",
      "handlebars": "^4.0.5",
      "jquery": "^2.2.3",
      "jquery-validation": "^1.15.0",
      "jquery.transit": "^0.9.12",
      "jstree": "^3.3.1",
      "lightbox2": "^2.8.2",
      "moment": "^2.13.0",
      "parse": "^1.8.5",
      "pikaday": "^1.4.0"
    },
    "browser": {
      "jquery": "./node_modules/admin-lte/plugins/jQuery/jQuery-2.2.0.min.js",
      "bootstrap": "./node_modules/admin-lte/bootstrap/js/bootstrap.min.js",
      "admin-lte": "./node_modules/admin-lte/dist/js/app.min.js"
    },
    "browserify": {
      "transform": [
        "browserify-shim"
      ]
    },
    "browserify-shim": {
      "jquery": "$",
      "bootstrap": {
        "depends": ["jquery:$"]
      },
      "admin-lte": {
        "depends": ["jquery:$", "bootstrap"]
      }
    }

With this config, the error changed -- AdminLTE is not defined at all. Here's the calling code:

    var AdminLTE = require('admin-lte');
    ...
    // this line causes the crash: AdminLTE is undefined -- neither accessing it as a member 
    // of $ nor globally works
    $.AdminLTE.layout.fix(); 

P.S.: there's a lot of dependent packages; I'm in the process of migrating a project from require.js to browserify, and most of these packages are not yet used in the migrated parts of the app.

**Comment by Maximization** [comment]
I don't see browserify-shim in your devDependencies but I'll assume it's installed. Other than that everything looks ok. Are you using Babel in the building process? If so you will have to include babelify to the shimming pipeline.

I did a console log on AdminLTE and it doesn't seem to export any methods. If you're trying to call anything on it it will crash. The way I use it in my project I only require it at the top so it will be bundled in the output file. What you're trying to do works though. Just make sure to import jQuery like this `var $ = require('jquery')` after AdminLTE, then you'll have access to `$.AdminLTE`.

PS: Not sure if it makes a difference but I have jquery installed as a dependency too

**Comment by drbarto** [comment]
browserify-shim is in the devDependencies, the second line of the code listing; I don't use Babel; jQuery is in my dependencies, and I `require` it before AdminLTE.

I did some more experiments; after `require('admin-lte)`, `$.AdminLTE` _is_ defined, but only contains one member `options`; it has no methods or other properties (like the `layout` property I was trying to access). Looks like AdminLTE is not imported correctly.

Thanks a lot for your help so far. If it works for you, then apparently my overall configuration is flawed in some way. If you have any further hints, I'd be very thankful; else I'll have to either abandon AdminLTE or browserify.

**Comment by Maximization** [comment]
Sorry I meant browserify. I required adminLTE before jQuery but switching them around doesn't seem to make a difference for me. Here's the console log I get for $.AdminLTE:

<img width="223" alt="__it_s_my_styl" src="https://cloud.githubusercontent.com/assets/9811142/15576847/2e172814-235a-11e6-8882-c392eb8c4fd0.png">

Get rid of this line
```
"jquery": "./node_modules/admin-lte/plugins/jQuery/jQuery-2.2.0.min.js",
```

It gave me a problem when I just inserted it. I remember I used to have it in there but I removed it at some point. Not sure what's going on under the hood, perhaps AdminLTE's jQuery plugin conflicts with the already installed jQuery module. Hopefully this solves your problem.
